File: Common/Src/ChinesePuzzle/Menu/MenuGridContainer.cpp
/**
 *  MenuGridContainer.cpp
 *  ChinesePuzzle
 *
 *  Created by Mathieu LEDRU on 01/11/11.
 *
 *  GPL License:
 *  Copyright (c) 2011, Mathieu LEDRU
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include "MenuGridContainer.h"

using namespace cocos2d;

MenuGridContainer::MenuGridContainer() :
container(NULL)
{
	
}

MenuGridContainer::~MenuGridContainer()
{
	CC_SAFE_RELEASE(container);
}

bool MenuGridContainer::initWithResolution(const char* resolution)
{
	if(!MenuBox::initWithResolution(resolution))
	{
		return false;
	}
	
	container = new MenuGrid();
	container->init();
	container->setAnchorPoint(ccp(0.5f, 0.5f));
	this->addChild(container);
	
	this->margin = CCSizeMake(0, 0);
	
	return true;
}

void MenuGridContainer::setItems(CCArray* items)
{
	container->setItems(items);
}

CCArray* MenuGridContainer::getItems()
{
	return container->getItems();
}

void MenuGridContainer::setGridSize(ccGridSize gridSize)
{
	container->setGridSize(gridSize);
}

ccGridSize MenuGridContainer::getGridSize()
{
	return container->getGridSize();
}

void MenuGridContainer::setMargin(cocos2d::CCSize margin)
{
	this->margin = margin;
	
	this->layout();
}

CCSize MenuGridContainer::getMargin()
{
	return this->margin;
}

void MenuGridContainer::setPage(int page)
{
	container->setPage(page);
}

int MenuGridContainer::getPage()
{
	return container->getPage();
}

void MenuGridContainer::setMinimumTouchLengthToSlide(cocos2d::CGFloat length)
{
	container->setMinimumTouchLengthToSlide(length);
}

CGFloat MenuGridContainer::getMinimumTouchLengthToSlide()
{
	return container->getMinimumTouchLengthToSlide();
}

void MenuGridContainer::setMinimumTouchLengthToChangePage(cocos2d::CGFloat length)
{
	container->setMinimumTouchLengthToChangePage(length);
}

CGFloat MenuGridContainer::getMinimumTouchLengthToChangePage()
{
	return container->getMinimumTouchLengthToChangePage();
}

void MenuGridContainer::layout()
{
	MenuBox::layout();
	
	CCSize size = this->getContentSize();
	
	if(container)
	{
		container->setPosition(ccp(size.width / 2, size.height / 2));
		container->setContentSize(CCSizeMake(size.width - 2 * margin.width, size.height - 2 * margin.height));
	}
}

bool MenuGridContainer::ccTouchBegan(CCTouch* pTouch, CCEvent* pEvent)
{
	if(!MenuBox::ccTouchBegan(pTouch, pEvent))
	{
		return false;
	}
	
	return container->ccTouchBegan(pTouch, pEvent);
}

void MenuGridContainer::ccTouchMoved(CCTouch* pTouch, CCEvent* pEvent)
{
	MenuBox::ccTouchMoved(pTouch, pEvent);
	
	container->ccTouchMoved(pTouch, pEvent);
}

void MenuGridContainer::ccTouchEnded(CCTouch* pTouch, CCEvent* pEvent)
{
	MenuBox::ccTouchEnded(pTouch, pEvent);
	
	container->ccTouchEnded(pTouch, pEvent);
}

void MenuGridContainer::ccTouchCancelled(CCTouch* pTouch, CCEvent* pEvent)
{
	MenuBox::ccTouchCancelled(pTouch, pEvent);
	
	container->ccTouchCancelled(pTouch, pEvent);
}

File: Common/Bin/Data/js/Menu/MenuGridContainer.js
/*
This file is part of the ChinesePuzzle package.

(c) Mathieu Ledru

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.
*/

cpz.MenuGridContainer = cpz.MenuBox.extend({
  _container: null,
  _gridSize: null,
  _margin: null,
  _page: null,
  _minimumTouchLengthToSlide: null,
  _minimumTouchLengthToChangePage: null,
  ctor: function() {
    return this._super();
  },
  initWithConf: function(conf) {
    if (!this._super(conf)) {
      return false;
    }
    this._container = new MenuGrid();
    this._container.init();
    this._container.setAnchorPoint(cc.p(0.5, 0.5));
    this.addChild(this._container);
    this._margin = cc.size(0, 0);
    return true;
  },
  getGridSize: function() {
    return this._container.getGridSize();
  },
  setGridSize: function(gridSize) {
    this._container.setGridSize(gridSize);
    return this;
  },
  getMargin: function() {
    return this._margin;
  },
  setMargin: function(margin) {
    this._margin = margin;
    this.layout();
    return this;
  },
  getPage: function() {
    return this._container.getPage();
  },
  setPage: function(page) {
    this._container.setPage(page);
    return this;
  },
  getMinimumTouchLengthToSlide: function() {
    return this._container.getMinimumTouchLengthToSlide();
  },
  setMinimumTouchLengthToSlide: function(length) {
    this._container.setMinimumTouchLengthToSlide(length);
    return this;
  },
  getMinimumTouchLengthToChangePage: function() {
    return this._container.getMinimumTouchLengthToChangePage();
  },
  setMinimumTouchLengthToChangePage: function(length) {
    this._container.setMinimumTouchLengthToChangePage(length);
    return this;
  },
  getItems: function() {
    return this._container.getItems();
  },
  setItems: function(items) {
    this._container.setItems(items);
    return this;
  },
  layout: function(anim) {
    var size;
    if (anim == null) {
      anim = true;
    }
    this._super(anim);
    size = this.getContentSize();
    if (this._container) {
      this._container.setPosition(cc.p(size.width / 2, size.height / 2));
      return this._container.setContentSize(cc.size(size.width - 2 * this._margin.width, size.height - 2 * this._margin.height));
    }
  },
  onTouchBegan: function(touch, event) {
    if (!this._super(touch, event)) {
      return false;
    }
    return this._container.ccTouchBegan(touch, event);
  },
  onTouchMoved: function(touch, event) {
    this._super(touch, event);
    return this._container.ccTouchMoved(touch, event);
  },
  onTouchEnded: function(touch, event) {
    this._super(touch, event);
    return this._container.ccTouchEnded(touch, event);
  },
  onTouchCancelled: function(touch, event) {
    this._super(touch, event);
    return this._container.ccTouchCancelled(touch, event);
  }
});
